platform/web/webmidi_driver.cpp

```diff
@@ -0,0 +1,86 @@
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "webmidi_driver.h"
+
+#include <emscripten.h>
+
+MIDIDriverWebMidi *MIDIDriverWebMidi::get_singleton() {
+	return static_cast<MIDIDriverWebMidi *>(MIDIDriver::get_singleton());
+}
+
+Error MIDIDriverWebMidi::open() {
+	Error error = godot_js_webmidi_open_midi_inputs(&MIDIDriverWebMidi::set_input_names_callback, &MIDIDriverWebMidi::on_midi_message);
+	if (error == ERR_UNAVAILABLE) {
+		ERR_PRINT("Web MIDI is not supported on this browser");
+	}
+	return error;
+}
+
+void MIDIDriverWebMidi::close() {
+	get_singleton()->connected_input_names.clear();
+	godot_js_webmidi_close_midi_inputs();
+}
+
+MIDIDriverWebMidi::~MIDIDriverWebMidi() {
+	close();
+}
+
+void MIDIDriverWebMidi::set_input_names_callback(int p_size, const char **p_input_names) {
+	Vector<String> input_names;
+	for (int i = 0; i < p_size; i++) {
+		input_names.append(String::utf8(p_input_names[i]));
+	}
+#ifdef PROXY_TO_PTHREAD_ENABLED
+	if (!Thread::is_main_thread()) {
+		callable_mp_static(MIDIDriverWebMidi::_set_input_names_callback).call_deferred(input_names);
+		return;
+	}
+#endif
+
+	_set_input_names_callback(input_names);
+}
+
+void MIDIDriverWebMidi::_set_input_names_callback(const Vector<String> &p_input_names) {
+	get_singleton()->connected_input_names.clear();
+	for (int i = 0; i < p_input_names.size(); i++) {
+		get_singleton()->connected_input_names.push_back(p_input_names[i]);
+	}
+}
+
+void MIDIDriverWebMidi::on_midi_message(int p_device_index, uint8_t p_status, const uint8_t *p_data, size_t p_data_len) {
+#ifdef PROXY_TO_PTHREAD_ENABLED
+	if (!Thread::is_main_thread()) {
+		MIDIDriver::send_event(p_device_index, p_status, p_data, p_data_len);
+		return;
+	}
+#endif
+	MIDIDriver::send_event(p_device_index, p_status, p_data, p_data_len);
```

Comment:
Isn't this the same function call twice?

Reply:
Yes, I meant to wrap it with `callable_mp_static` like in the other callback. However, I just added this following the pattern of the other static callbacks I've seen and am not sure how to ensure this branch is working properly.